tarxx-example.cpp

```diff
@@ -23,48 +23,191 @@
+    return tar_files_in(tar, input_files);
+}
+
+static int tar_files_in_file_out(const std::string& output_file,
+                                 const std::vector<std::string>& input_files,
+                                 const tarxx::tarfile::compression_mode& compression_mode)
+{
+    tarxx::tarfile tar(output_file, compression_mode);
+    return tar_files_in(tar, input_files);
+}
+
+static int tar_stream_in(tarxx::tarfile& tar)
 {
-    if (argc < 2) return 1;
-    tarxx::tarfile tar(argv[1] + std::string("_stream.tar"));
     if (!tar.is_open()) return 2;
-    for (unsigned i = 2; i < argc; ++i) {
-        tar.add_file_streaming();
-        std::ifstream infile(argv[i], std::ios::binary);
-        std::array<char, 1466> buf {};
-        std::streamsize file_size = 0;
-
-        while (infile.good()) {
-            infile.read(buf.data(), buf.size());
-            const auto read = infile.gcount();
-            file_size += read;
-            tar.add_file_streaming_data(buf.data(), read);
-        }
 
-        tar.stream_file_complete(argv[i], 0777, 0, 0, file_size, 0);
+    std::array<char, 1> input_buff {};
+    long long total_size = 0;
+    tar.add_file_streaming();
+
+    while (std::cin.read(input_buff.data(), input_buff.size())) {
+        const auto read_bytes = std::cin.gcount();
+        total_size += read_bytes;
+        tar.add_file_streaming_data(input_buff.data(), read_bytes);
     }
+
+    std::time_t result = std::time(nullptr);
+
+#if __linux
+    tar.stream_file_complete("stdin", 0777, getuid(), getgid(), total_size, result);
+#endif
     return 0;
 }
 
-int main(const int argc, const char* const* const argv)
+static int tar_stream_in_file_out(const std::string& output_file,
+                                  const tarxx::tarfile::compression_mode& compression_mode)
+{
+    tarxx::tarfile tar(output_file, compression_mode);
+    return tar_stream_in(tar);
+}
+
+static int tar_stream_in_stream_out(std::ostream& os,
+                                    const tarxx::tarfile::compression_mode& compression_mode)
+{
+    tarxx::tarfile::callback_t cb = [&](const tarxx::block_t& block, const size_t size) {
+        os.write(block.data(), size);
+    };
+
+    tarxx::tarfile tar(std::move(cb), compression_mode);
+    return tar_stream_in(tar);
+}
+
+static bool std_out_redirected()
+{
+    return isatty(fileno(stdout)) == 0;
+}
+
+static bool std_in_redirected()
 {
-    auto returnValue = tar_file_mode(argc, argv);
-    if (returnValue != 0) return returnValue;
+    return isatty(fileno(stdin)) == 0;
+}
+
+
+int main(const int argc, char* const* const argv)
+{
+    int opt;
+    auto compress = false;
+    auto create = false;
+    std::string filename;
+
+    while ((opt = getopt(argc, argv, "kf:c")) != -1) {
+        switch (opt) {
+            case 'c':
+                create = true;
+                break;
+            case 'k':
+                compress = true;
+                break;
```

Comment:
let's give that option (and provide the processing of it below) only if the user actually compiles with a compression method enabled (see comment in `CMakeLists`).